server/aesdsocket.c

```diff
@@ -161,32 +197,220 @@ void runAsDaemon()
         close(extf);
 }
 
+// receives and sends throught the sockets
+void *socketHandler(void *data)
+{
+	
+	char buffer[1024];
+	
+	threads_data *thread_data = (threads_data*)data;
+	
+	rdBuff = malloc(sizeof(char));
+	
+	int pe = pthread_mutex_lock(&mutex);
+	if(pe != 0){
+		LOG_ERROR("Error in mutex lock: %s",strerror(errno));
+		closeAll();
+		exit(EXIT_FAILURE);
+	}
+	
+	
+	// Receive and write to WRITE_FILE
+	int fd =open(WRITE_FILE,O_WRONLY);
+	if(fd < 0){
+		LOG_ERROR("Error opening %s",WRITE_FILE);
+		closeAll();
+		exit(1);
+	}
+	
+        lseek(fd,0,SEEK_END);
+        unsigned long index = 0;
+	while(1){
+	
+		ssize_t rBytes = recv(thread_data->clientfd, buffer, sizeof(buffer), 0);
+		if(rBytes == -1){
+			LOG_ERROR("recv():%s", strerror(errno));
+			closeAll();
+			exit(1);
+		}
+		rdBuff = realloc(rdBuff,index+rBytes);
+		memcpy(rdBuff+index,buffer,rBytes);
+		index += rBytes;
+		totalBytes += rBytes;
+		if(buffer[rBytes-1] == '\n'){
+			
+			if(write(fd, rdBuff, index) == -1){
+				LOG_ERROR("write():%s", strerror(errno));
+				closeAll();
+				exit(1);
+			}
+			break;
+		}
+	}
+	close(fd);	
+	free(rdBuff);
+	rdBuff=NULL;	
+	
+	// Send complete data from WRITE_FILE to the client
+	fd =open(WRITE_FILE,O_RDONLY);
+	if(fd < 0){
+		LOG_ERROR("Error opening %s",WRITE_FILE);
+		closeAll();
+		exit(1);
+	}
+	ssize_t sendBytes = totalBytes;
+	rdBuff = realloc(rdBuff,sendBytes);
+	lseek(fd,0,SEEK_SET);
+	while(1){		
+		
+		int rBytes = read(fd, rdBuff,sendBytes);
+		if(rBytes == -1){
+			LOG_ERROR("read():%s", strerror(errno));
+			closeAll();
+			exit(1);
+		}
+		if(send(thread_data->clientfd, rdBuff, rBytes, 0) == -1){
+			LOG_ERROR("send():%s", strerror(errno));
+			closeAll();
+			exit(1);
+		}
+		sendBytes -= rBytes;
+		if(sendBytes == 0)
+			break;
+	}
+	free(rdBuff);
+	rdBuff=NULL;
+	close(fd);
+	thread_data->thread_status = T_COMPLETE;
+	
+	pe = pthread_mutex_unlock(&mutex);
```

Comment:
@ajsanthosh14  Do you need the lock for entire block of code? Are you locking the code here or the shared resource?